### ISSUE TITLE
Upcoming milestones

### ISSUE THREAD

**Issue by AdityaKane2001** [AdityaKane2001/regnety#2]
Hello @sayakpaul and @MorganR ,

Following are the milestones for the first couple of weeks.  

Data preprocessing and input pipeline implementation and testing (Weeks 1 and 2)
1.  **Augmentations to be applied** 
2. **TFRecords config**
3. **Testing speed and correctness of input pipeline**
4. GCS bucket config
5. Configuration of GCP VM and TPU 

I believe the first three of them are more or less completed, and we are now ironing out the finer details. They will be formally completed when the PR is merged. The latter two will be completed as soon as I have the GCP access, and should ideally be done within three days at the most. 
This brings us to the next couple of weeks and the milestones are:

Actual model scripting and testing (Weeks 3 and 4)
1. Implementation of base level classes (Block, Stage, etc.) 
2. Implementation of 4 RegNetY variations
3. Testing the model(s) on small datasets (Imagenette)

Is it okay if I start the development of the model script on a separate branch? The current open PR can be merged as soon as it's ready. I believe that there are no more changes to the core functionality of the `imagenet.py` and `tfrecords_utils.py` scripts and they can be assumed frozen for future development. 

Please share your thoughts on the same. 
Thanks.

**Comment by sayakpaul** [comment]
Sounds great and thanks for being so comprehensive about your goals. Should we maintain kanban cards by using the Projects feature of GitHub? @MorganR WDYT? 

For the first three ones, I am game except for a small ask. I'd really like to see a notebook that loads the TFRecords, constructs the data input pipeline, and visualizes a few image batches. 4 and 5 are trivial and I am sure Morgan and I can get you up to speed with those. So, allocating 1/2 days are sufficient IMO. 

> The current open PR can be merged as soon as it's ready. I believe that there are no more changes to the core functionality of the imagenet.py and tfrecords_utils.py scripts and they can be assumed frozen for future development.

I agree. We will only be able to confirm this once we are able to set things up on GCP. For the time being, let's test things with ImageNette because it provides a reasonable representation of what we'd be dealing with. Does that make sense? 

> Is it okay if I start the development of the model script on a separate branch?

I am okay with this. @MorganR WDYT?

**Comment by AdityaKane2001** [comment]
>  I'd really like to see a notebook that loads the TFRecords, constructs the data input pipeline, and visualizes a few image batches. 

Yes, totally. The notebook's almost ready, just making some minor changes.

>I agree. We will only be able to confirm this once we are able to set things up on GCP. For the time being, let's test things with ImageNette because it provides a reasonable representation of what we'd be dealing with. Does that make sense?

Yes, sounds good.

**Comment by AdityaKane2001** [comment]
@sayakpaul [Here's ](https://colab.research.google.com/gist/AdityaKane2001/4b4e32d45cb95f46bba59e43e35c63f1/regnety_dataloader.ipynb?authuser=1) the demo notebook. Hope it's as expected.

**Comment by sayakpaul** [comment]
> Hope it's as expected.

No actually. I would like to parse the TFRecords created and then apply the preprocessing functions on them. Then I would like to visualize a couple of batches, and then maybe inspect the latency.

**Comment by AdityaKane2001** [comment]
If I understand correctly, you want to see the effect of preprocessing functions on the images. For that, 10 original and their augmented counterparts are saved in Colab's `/content/` directory after the execution of the two cells in the notebook. 

About the latency, I'll look up the profiler and get back to you.

**Comment by sayakpaul** [comment]
> If I understand correctly, you want to see the effect of preprocessing functions on the images. For that, 10 original and their augmented counterparts are saved in Colab's /content/ directory after the execution of the two cells in the notebook.

it's not very self-contained then. It assumes that the viewer will the notebook entirely and then be able to see the results. It's a bit hectic for review purposes. Refer to [this notebook](https://github.com/sayakpaul/TF-2.0-Hacks/blob/master/Cats_vs_Dogs_TFRecords.ipynb) and you'd get an idea of what I meant.

**Comment by AdityaKane2001** [comment]
@sayakpaul 
1. [Here's](https://colab.research.google.com/gist/AdityaKane2001/90e20c91e708abfe5d4bdfb59558a409/regnety_dataloader.ipynb?authuser=1) the notebook. Please tell me if it needs any changes. Please do not execute the cells, as it may give an error as I have one hot encoded the labels and batched the dataset in a latter commit.
2.  I tried the profiler but it was stuck at "Loading data" for about an hour. Found the same issue with the official tutorial notebook. 
3. There are some very minor changes in `imagenet.py` and `tfrecords_utils.py` (of a couple of lines each). I will include them in the next PR, along with the model code. 

Thanks
/cc @MorganR

**Comment by sayakpaul** [comment]
Would prefer to review the data related changes first. So please create a PR to include those and have us review those. 

Regarding the loader being stuck, could you open an issue for that? Or even post your doubt in the TensorFlow Discuss Forum?

I will take a look into the notebook tomorrow.

**Comment by sayakpaul** [comment]
The notebook looks more comprehensive now. Thank you. 

For this one, 

![image](https://user-images.githubusercontent.com/22957388/123061637-7238ba00-d429-11eb-997b-c12fcbc1c340.png)

Should we experiment with different scales to do the cropping? But I think it's kind of common to have these kinds of noisy samples to regularize the training sometimes.

**Comment by AdityaKane2001** [comment]
@sayakpaul 
[Here's](https://colab.research.google.com/gist/AdityaKane2001/7a8ebef9038d28d19c4f505628b73637/regnety_dataloader.ipynb?authuser=1#scrollTo=h_a2JWIR2oXn) the updated notebook.

But I'd like to point out that this notebook cannot be executed, as we convert the examples such that they are model-ready, which means removing all other attributes like `filename`,`synset`,etc. So if the objective is to make a tutorial notebook, then I can make that separately. 

But if we want this notebook for our internal debugging, then I think this one might be what we need. By that I mean a notebook  where we can see the actual augmentations and compare them to the original images. Please correct me if I am wrong.

**Comment by sayakpaul** [comment]
> But if we want this notebook for our internal debugging, then I think this one might be what we need. By that I mean a notebook where we can see the actual augmentations and compare them to the original images. Please correct me if I am wrong.

Works. We can consider including this notebook (with extensive warnings about its execution) when we finally release the implementation. I will take a look at the notebook later today/tomorrow and get back.

**Comment by sayakpaul** [comment]
For sections like "Display help and run script to make TFRecords", I think it would also make sense to include one/two lines of description about what the script is doing internally. For example:

> `main.py` is responsible for loading images from disk, resizing them to a uniform shape, ...

This is helpful for self-revision purposes too. 

The description under `Instantiate imagenet.ImageNet class.` this one is just spot on! Great job. c

Additionally, consider the comments shared [here](https://github.com/AdityaKane2001/regnety/issues/2#issuecomment-866631924) with respect to resizing and cropping.

**Comment by AdityaKane2001** [comment]
> But I think it's kind of common to have these kinds of noisy samples to regularize the training sometimes.

I agree. Many models do implement this, but it seemed a bit counter-intuitive.

>Or maybe, we could first resize to the larger dimension (larger between height and the width) and then apply cropping.

I don't think this will help a lot, as the crop can still be very small. Also then the image will be substantially distorted.

I have updated the notebook. Actually, I reused the docstrings instead of writing something new. I did the same with "Display help and run script to make TFRecords".

**Comment by sayakpaul** [comment]
What about this?

> Should we experiment with different scales to do the cropping?

**Comment by AdityaKane2001** [comment]
Different scales as in something other than [0.75, 1.33] right? If so, I'm okay with that.

I had made a couple of docs earlier to maintain such extra experiments (apart from the stock ones mentioned in paper/ official repo). Please see the docs below:

1. [AdityaKane_gsoc_stretch_goals](https://docs.google.com/document/d/1Y-Km5EgzCnCyO1fXzhd1ycVR0J5O3dIfwnhTtUGwhEs/edit?usp=sharing)
2. [RegNet training and augmentations summary ](https://docs.google.com/document/d/1kgOW4td8NBk3mrvQSj_Rrx0drbY37vekjcAoXVxrF64/edit?usp=sharing)


We could maintain any additional thoughts or experiments in these docs. Please feel free to add any points to these docs as you wish.

**Comment by sayakpaul** [comment]
For now, I would just vary the scale to gauge the possibilities. Hoe about narrowing down the scale interval?

**Comment by AdityaKane2001** [comment]
@sayakpaul @MorganR 

I wanted to talk a bit about the upcoming weeks. 

Following are the objectives for weeks 5, 6 and 7 from the doc:
1. Training of 4 models on GCP
2. Tweaking and retraining models (if necessary) for SOTA results
3. Deploying SavedModel to TensorFlow Hub and checkpoint files to GitHub

These are my thoughts on the same: 

In the upcoming weeks, the plan was to train the models and upload them to TFHub and GitHub. That's not possible now, as we've not fixed the model script yet and also the credits issue. It would be really great if you could take a look at the model, so the scripts can be ready to go when the credits are available. We'll also require some time to fix the learning scheme we want to implement for training.

I understand that I'm rushing a bit, but it's just because I'm not sure how long training and refining the models will take. 

Please share your thoughts regarding this. 

Thanks.

**Comment by sayakpaul** [comment]
IMO, model scripts won't take nearly as long as the data input pipeline. If we three can agree on a fixed date to complete the model scripts I think it should still be achievable. 

We can review the model scripts but switching back and forth between different feedback is harmful to the cognitive load is what I think. @MorganR WDYT? 

But on a serious note, if the credits' issue is not resolved then we need to expedite things a bit. 

> We'll also require some time to fix the learning scheme we want to implement for training.

Let's try to rely on what the authors follow in the paper. We can even start with the classic ResNet50 training configurations and then take it from there. Refinement of the training configuration will only be clear once we have some performance numbers to compare to.

**Comment by AdityaKane2001** [comment]
>  model scripts won't take nearly as long as the data input pipeline. If we three can agree on a fixed date to complete the model scripts I think it should still be achievable.

That would be great. We'll discuss this in the next meet if that works.

> Let's try to rely on what the authors follow in the paper.

Sure, also there's a lot on this in the official repo.

**Comment by sayakpaul** [comment]
In that case, starting with the known heuristics like ResNet50 training configs along with the suggestions from [here](https://arxiv.org/abs/1706.02677) would be a good place to start.

**Comment by AdityaKane2001** [comment]
Please see #7 for continuation.
